src/Covenant/Internal/Unification.hs

```diff
@@ -42,36 +52,93 @@ data TypeAppError
   | -- | We were given too many arguments.
     ExcessArgs (CompT Renamed) (Vector (Maybe (ValT Renamed)))
   | -- | We weren't given enough arguments.
-    InsufficientArgs (CompT Renamed)
+    InsufficientArgs Int (CompT Renamed) [Maybe (ValT Renamed)]
   | -- | The expected type (first field) and actual type (second field) do not
     -- unify.
     DoesNotUnify (ValT Renamed) (ValT Renamed)
+  | -- | No datatype info associated with requested TyName
+    NoDatatypeInfo TyName
+  | -- | No BB form. The only datatypes which should lack one are those isomorphic to `Void`
+    NoBBForm TyName
+  | -- | Datatype renaming failed. NOTE/REVIEW: Using `DatatypeInfo AbstractTy` here lets use use the same
+    --   reader context as ASGBuilder, which seems like the simplest change possible
+    --   I guess actually this means that renaming the BB form of a datatype failed
+    DatatypeInfoRenameFailed TyName RenameError
+  | -- | Something happened that definitely should not have. For right now, this means: The BB form of a datatype isn't a thunk
+    --   (but it might be useful to keep this around as a catchall for things that really shouldn't happen)
+    ImpossibleHappened Text
```

Comment:
Need `since`s, since they're new constructors.

Also, the NOTE shouldn't be in the Haddocks.